File: CoinChangeII/change.py
#!/usr/bin/env python3

class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        prev_dp = [1] + [0]*(amount)
        for i in range(len(coins)-1, -1, -1):
            dp = [1] + [0]*(amount)
            for amt in range(1, amount+1):
                dp[amt] = prev_dp[amt]
                if coins[i] <= amt:
                    dp[amt] += dp[amt-coins[i]]
            prev_dp = dp
        return dp[amount]

File: CoinChangeII/change.cpp
#include <vector>

class Solution {
public:
    int change(int amount, vector<int>& coins) {
        std::vector<int> prev(amount+1, 0);
        for (int i = coins.size()-1; i >= 0; --i){
            std::vector<int> dp(amount+1, 0);
            dp[0] = 1;
            for (int j = 1; j < amount+1; ++j){
                dp[j] = prev[j];
                if (j >= coins[i])
                    dp[j] += dp[j-coins[i]];
            }
            prev = dp;
        }
        return prev[amount];
    }
};

